### PR TITLE
api: support inverting matches in rate limit

### LINKED ISSUE #2193: AND NOT style matches for RateLimiting
*Description*:
>Describe the desired behavior, what scenario it enables and how it
would be used.

Consider this config
```
cat <<EOF | kubectl apply -f -
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: jwt-backend
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: backend
  jwt:
    providers:
    - name: example
      remoteJWKS:
        uri: https://test.eu.auth0.com/.well-known/jwks.json
      claimToHeaders:
      - claim: email
        header: x-user-email
---
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: BackendTrafficPolicy 
metadata:
  name: policy-httproute
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: backend
    namespace: envoy
  rateLimit:
    type: Global
    global:
      rules:
      - clientSelectors:
        - headers:
          - name: x-user-email
            value: myemail@gmail.com
        limit:
          requests: 10
          unit: Minute
    - clientSelectors:
      - headers:
        - type: Distinct
          name: x-claim-email
      limit:
        requests: 5
        unit: Minute
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: backend
spec:
  parentRefs:
    - name: eg
      namespace: envoy
  hostnames:
    - "www.example.com"
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: backend
          port: 3000
          weight: 1
      matches:
        - path:
            type: PathPrefix
            value: /
EOF
```

I want to allow [myemail@gmail.com](mailto:myemail@gmail.com) user to 10 req/min but all other user 5 req/min.
But with this config [myemail@gmail.com](mailto:myemail@gmail.com) user is getting 429 after 5 requests.

Relates to https://envoyproxy.slack.com/archives/C03E6NHLESV/p1699784524023299

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

### PR DESCRIPTION
**Pull request by rudrakhp**
**What type of PR is this?**

Added an `InvertMatch` field in `SourceMatch` and `HeaderMatch` types.

**What this PR does / why we need it**:

Extends the API to support inverted matches, which enables AND NOT style matches for rate limiting.

**Which issue(s) this PR fixes**:

Ref #2193

### CODE REVIEW

File: api/v1alpha1/ratelimit_types.go [hunk 1]
```diff
@@ -138,6 +138,12 @@ type SourceMatch struct {
 	// +kubebuilder:validation:MinLength=1
 	// +kubebuilder:validation:MaxLength=256
 	Value string `json:"value"`
+
+	// InvertMatch specifies whether the value match result will be inverted.
+	//
+	// +optional
+	// +kubebuilder:default=false
```
Comment (shawnh2) [inline api/v1alpha1/ratelimit_types.go:145]: need a `// +notImplementedHide` comment, and remove it once it has been implemented

File: api/v1alpha1/ratelimit_types.go [hunk 2]
```diff
@@ -138,6 +138,12 @@ type SourceMatch struct {
...
+	InvertMatch bool `json:"invertMatch,omitempty"`
```
Comment (shawnh2) [inline api/v1alpha1/ratelimit_types.go:146]: Prefer just `Invert`

File: api/v1alpha1/ratelimit_types.go [hunk 3]
```diff
@@ -161,6 +167,14 @@ type HeaderMatch struct { // TODO: zhaohuabing this type could be replaced with
...
 	// +kubebuilder:validation:MaxLength=1024
 	Value *string `json:"value,omitempty"`
+
+	// InvertMatch specifies whether the value match result will be inverted.
+	// Do not set this field when Type="Distinct", implying matching on any/all unique
+	// values within the header.
+	//
+	// +optional
+	// +kubebuilder:default=false
+	InvertMatch bool `json:"invertMatch,omitempty"`
```
Comment (shawnh2) [inline api/v1alpha1/ratelimit_types.go:177]: ditto

File: api/v1alpha1/ratelimit_types.go (hunk 2 above)
Comment (shawnh2) [reply api/v1alpha1/ratelimit_types.go:146]: Also should be a pointer since it's an optional type

Comment (rudrakhp) [reply api/v1alpha1/ratelimit_types.go:146]: @shawnh2 I am not sure what `optional` means here, want this param to be optional for user to set but default interpretation should be `false`, implementation shouldn't have to check a for nil pointer. Let me know what you think!

File: api/v1alpha1/ratelimit_types.go [hunk 4]
```diff
@@ -161,6 +168,15 @@ type HeaderMatch struct { // TODO: zhaohuabing this type could be replaced with
 	// +optional
 	// +kubebuilder:validation:MaxLength=1024
 	Value *string `json:"value,omitempty"`
+
+	// Invert specifies whether the value match result will be inverted.
+	// Do not set this field when Type="Distinct", implying matching on any/all unique
```
Comment (zhaohuabing) [inline api/v1alpha1/ratelimit_types.go:173]: Can we add a CEL validation for this?

File: api/v1alpha1/ratelimit_types.go [hunk 5]
```diff
@@ -161,6 +168,15 @@ type HeaderMatch struct { // TODO: zhaohuabing this type could be replaced with
...
+	// values within the header.
+	//
+	// +optional
+	// +kubebuilder:default=false
+	// +notImplementedHide
+	Invert bool `json:"invertMatch,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/ratelimit_types.go:179]: should this be `Invert` or `InvertMatch` ? @envoyproxy/gateway-maintainers @envoyproxy/gateway-reviewers

Comment (zhaohuabing) [reply api/v1alpha1/ratelimit_types.go:179]: Slightly prefer `InvertMatch` for clarity, but it's not a strong opinion.

Comment (shawnh2) [reply api/v1alpha1/ratelimit_types.go:179]: Prefer `Invert` just keep it simple

Comment (arkodg) [reply api/v1alpha1/ratelimit_types.go:179]: +1 ing `Invert`

File: api/v1alpha1/ratelimit_types.go [hunk 6]
```diff
@@ -138,6 +138,13 @@ type SourceMatch struct {
 	// +kubebuilder:validation:MinLength=1
 	// +kubebuilder:validation:MaxLength=256
 	Value string `json:"value"`
+
+	// Invert specifies whether the value match result will be inverted.
+	//
+	// +optional
+	// +kubebuilder:default=false
+	// +notImplementedHide
+	Invert bool `json:"invertMatch,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/ratelimit_types.go:147]: can this be made into a `*bool` ? to ensure consistency across all optional fields in the API

File: api/v1alpha1/ratelimit_types.go (hunk 4 above)
Comment (arkodg) [reply api/v1alpha1/ratelimit_types.go:173]: hey @zhaohuabing can you provide some hints/examples for @rudrakhp to try out ?
these CEL validations can be validated in https://github.com/envoyproxy/gateway/blob/main/test/cel-validation/backendtrafficpolicy_test.go

**Review by arkodg — APPROVED** [review]
LGTM thanks !

should be fine to add the CEL validation as a follow up

**Review by shawnh2 — APPROVED** [review]
LGTM

**Pull request merged**